### PR TITLE
Remove use of assert_throws in idlharness.js

### PR DESCRIPTION
**Pull request by bzbarsky**

### CODE REVIEW

**Comment by bzbarsky** [comment]
@jgraham @stephenmcgruer The failing bit is Firefox nightly stability, but this is green in Firefox CI on web platform tests; see https://treeherder.mozilla.org/#/jobs?repo=try&revision=c38267ce319fc9a8c49724e71ad3a00f4d99b59e

**Comment by stephenmcgruer** [comment]
I don't think I know enough JS to review this :/. Can you point me at some documentation or a further description of what problem you're solving with `globalOf`?

This may be worth running on a [trigger branch](https://web-platform-tests.org/running-tests/from-ci.html?highlight=trigger), to see if there are any regressions.

**Comment by bzbarsky** [comment]
> Can you point me at some documentation or a further description of what problem you're solving with globalOf

Sure.  The assert_throws_js method checks not only that the right sort of exception is thrown but also that it's thrown from the right global.  That is, it checks that the `.constructor` of the thrown exception is the object passed in as the first arg.

Now in most tests this is not a problem: you examine which function you are calling, decide what global it lives in, and use that global to fetch the object.  But in idlharness you could be dealing with objects from various globals; for example some of the Document interface tests pass in both the main document for the page and a subframe document.  So what globalOf is doing is taking a JS function being tested, and determining its global, so we can assert that the function threw exceptions whose constructor is in fact that global.

I'll try the trigger branch thing.

**Comment by bzbarsky** [comment]
OK, I'm clearly doing something wrong.  When I do:
```
git push --force-with-lease git@github.com:web-platform-tests/wpt.git idlharness-assert-throws:triggers/firefox_nightly
```
I get:
```
 ! [rejected]              idlharness-assert-throws -> triggers/firefox_nightly (stale info)
```
Do I need to pull the triggers/firefox_nightly branch first and cherrypick my changeset onto it?  I would have thought that --force-with-lease would mean that's not needed....  (Its also possible I don't have the right permissions, of course.)

**Comment by stephenmcgruer** [comment]
> Its also possible I don't have the right permissions, of course

This appears to be the case. I have pushed this commit to `triggers/chrome_dev` and `triggers/firefox_nightly`. I will keep an eye on wpt.fyi/runs for this commit.

In the meantime, thank you for the information on how the change works - I will digest and then review! :)

File: resources/idlharness.js [hunk 1]
```diff
@@ -84,10 +84,26 @@ function minOverloadLength(overloads)
     .reduce(function(m, n) { return Math.min(m, n); });
 }
 
+// A helper to get the global of a Function object.  This is needed to determine
+// which global exceptions the function throws will come from.
+function globalOf(func)
+{
+    var global;
```
Comment (stephenmcgruer) [inline resources/idlharness.js]: Not sure if this works in JS, but could we just do:

```
try {
  return func.constructor("return this;")();
} catch (e) {}
return self;
```

File: resources/idlharness.js [hunk 2]
```diff
@@ -84,10 +84,26 @@ function minOverloadLength(overloads)
...
+    try {
+        // Use the fact that .constructor for a Function object is the Function
+        // constructor, which can be used to mint a new function in the right
+        // global.
+        global = func.constructor("return this;")();
+    } catch (e) {
+        // If it fails, just fall back to 'self'.
```
Comment (stephenmcgruer) [inline resources/idlharness.js]: Would be useful to explain what it means if this fails. Was 'func' not a function then?

File: resources/idlharness.js [hunk 3]
```diff
@@ -2160,7 +2176,17 @@ IdlInterface.prototype.test_immutable_prototype = function(type, obj)
             } catch (err) {}
         });
 
-        assert_throws(new TypeError(), function() {
+        var setter;
```
Comment (stephenmcgruer) [inline resources/idlharness.js]: Can we use modern `let` (or `const`) rather than `var` (unless they have to be global for some reason?)

Comment (stephenmcgruer) [inline resources/idlharness.js]: This block is complex enough that I think it deserves a quick comment as to why it needs to exist.

File: resources/idlharness.js (hunk 1 above)
Comment (bzbarsky) [reply resources/idlharness.js]: It depends on how defensive we're trying to be.  For example, that return statement inside the try/catch block could return `undefined` or something, depending on how off we are on what `func` is.  To be even more defensive, we could test that `global` is actually an `Object` and return `global.self || self`.  In practive, hopefully it all doesn't matter, so maybe we should just do the simple thing.

File: resources/idlharness.js (hunk 2 above)
Comment (bzbarsky) [reply resources/idlharness.js]: Or a function with a weird `.constructor` defined on it, or with a weird proto chain....  I can add a comment,

File: resources/idlharness.js (hunk 3 above)
Comment (bzbarsky) [reply resources/idlharness.js]: We can use `let`, sure.  At one point we were avoiding that in wpt due to poor support, but I think it's good to put that behind us.  ;)

Comment (bzbarsky) [reply resources/idlharness.js]: Yep, will add.

**Review by stephenmcgruer — APPROVED** [review]
Thanks! I checked the trigger runs for the previous commit and they looked fine, so I think this is good to go :)

File: resources/idlharness.js [hunk 4]
```diff
@@ -2151,7 +2168,7 @@ IdlInterface.prototype.test_immutable_prototype = function(type, obj)
         var newValue = Object.create(null);
 
         t.add_cleanup(function() {
-            var setter = Object.getOwnPropertyDescriptor(
+            let setter = Object.getOwnPropertyDescriptor(
```
[the remark this answers is not in the document]
Comment (bzbarsky) [reply resources/idlharness.js]: Just lost track of which vars I had introduced and which were preexisting.

File: resources/idlharness.js [hunk 5]
```diff
@@ -2160,7 +2177,22 @@ IdlInterface.prototype.test_immutable_prototype = function(type, obj)
...
 
-        assert_throws(new TypeError(), function() {
+        // We need to find the actual setter for the '__proto__' property, so we
+        // can determine the right global for it.  Walk up the prototype chain
+        // looking for that property until we find it.
+        let setter;
+        {
+            let cur = obj;
+            while (cur) {
+                var desc = Object.getOwnPropertyDescriptor(cur, "__proto__");
```
[the remark this answers is not in the document]
Comment (bzbarsky) [reply resources/idlharness.js]: Good catch.

File: resources/idlharness.js [hunk 6]
```diff
@@ -84,10 +84,27 @@ function minOverloadLength(overloads)
     .reduce(function(m, n) { return Math.min(m, n); });
 }
 
+// A helper to get the global of a Function object.  This is needed to determine
+// which global exceptions the function throws will come from.
```
Comment (stephenmcgruer) [inline resources/idlharness.js]: Fun irrelevant fact, 'This is needed to determine which global exceptions the function throws will come from.' is a [garden path sentence](https://en.wikipedia.org/wiki/Garden-path_sentence). The 'global' is read as an adjective until you reach 'will come from', at which point it needs to be reparsed as a 'noun' :D.

**Pull request merged**